snuba/datasets/entity.py

```diff
@@ -0,0 +1,101 @@
+from typing import Any, Mapping, Optional, Sequence, Tuple
+
+from snuba.clickhouse.columns import ColumnSet
+from snuba.clickhouse.escaping import escape_identifier
+from snuba.datasets.plans.query_plan import ClickhouseQueryPlanBuilder
+from snuba.datasets.storage import Storage, WritableStorage, WritableTableStorage
+from snuba.query.extensions import QueryExtension
+from snuba.query.logical import Query
+from snuba.query.parsing import ParsingContext
+from snuba.query.processors import QueryProcessor
+from snuba.util import qualified_column
+
+
+class Entity(object):
+    def __init__(
+        self,
+        *,
+        storages: Sequence[Storage],
+        query_plan_builder: ClickhouseQueryPlanBuilder,
+        abstract_column_set: ColumnSet,
+        writable_storage: Optional[WritableStorage],
+    ) -> None:
+        self.__storages = storages
+        self.__query_plan_builder = query_plan_builder
+        self.__writable_storage = writable_storage
+        # TODO: This data model will change as we add more functionality
+        self.__data_model = abstract_column_set
+
+    def get_extensions(self) -> Mapping[str, QueryExtension]:
+        """
+        Returns the extensions for this entity.
+        Every extension comes as an instance of QueryExtension.
+        The schema tells Snuba how to parse the query.
+        The processor actually does query processing for this extension.
+        """
+        raise NotImplementedError("entity does not support queries")
+
+    def get_query_processors(self) -> Sequence[QueryProcessor]:
+        """
+        Returns a series of transformation functions (in the form of QueryProcessor objects)
+        that are applied to queries after parsing and before running them on the storage.
+        These are applied in sequence in the same order as they are defined and are supposed
+        to be stateless.
+        """
+        return []
+
+    def get_data_model(self) -> ColumnSet:
+        """
+        Now the data model is flat so this is just a simple ColumnSet object. We can expand this
+        to also include relationships between entities.
+        """
+        return self.__data_model
+
+    def get_query_plan_builder(self) -> ClickhouseQueryPlanBuilder:
```

Comment:
Back to my previous comment about the plan builder. Now that we start putting some bricks in place and subdividing the responsibilities, we should look into how the execution of a query would work both in a single entity case and in a multiple entity case and have a rough idea specifically of the multi entity case (join).

Said that, I would be ok in making a step in this direction with this PR that addresses only the single entity query (thus moves the query plan builder into the entity) as an intermediate step to focus on the query plan in tthe following step.

Reply:
I added a shim to the dataset to call the entity, so the function is there and can be expanded in the future.